enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,316 @@
+As master node x being restarted the VIP is owned by another node.
+
+- T0: kube-apiserver returns OK on /readyz
+- T1: Keepalived static pod starts running, VIP should remain on node != master-x
+- T2: (<= T0 + 2Sec) HAProxy (on the node holds the VIP) will mark master-x backend as healthy, and should send new connections also towards master-x.
+- T3: (= T2 + 30Sec ) HAProxy pod starts running on master-x, that shouldn't affect API traffic since VIP remains on node != master-x 
+
+##### Restart of Master node - external LB,DNS
+- T0: kube-apiserver returns OK on /readyz
+- T1: (= T0 + Y Sec , Y=external LB backend health detection time) external LB should mark master-x backend as ready, and should send new connections also towards master-x.
+
+##### Disaster recovery - self-hosted LB
+- T0: kube-apiserver recovered
+- T1: Keepalived static starts running, since this master is the only node running Keepalived at this point, it will hold the VIP.
+- T2: (= T1 + ~30 Sec) HAProxy starts running, install iptable rule that redirects traffic to LB
+
+##### Disaster recovery - external LB,DNS
+- T0: kube-apiserver recovered
+- T1: kube-apiserver returns OK on /readyz
+- T2: (= T1 + Y Sec , Y=external LB backend health detection time) External LB should mark master-x backend as ready, and should send new connections towards master-x.
+
+### cluster-hosted-net-services operator
+The `cluster-hosted-net-services-operator` is a new operator whose operand is a controller for the cluster hosted DNS, LB and VIP services.
+
+### Standard SLO Behavior
+
+The CHNSO is expected to follow the standard behavior of SLO, additionally, since the CHNSO is applicable only to clusters running on on-prem ( 'baremetal', 'openstack', 'vsphere' or 'ovirt') infrastructure,
+It should support also the expected behaviors for "not in use" SLOs similar to the [CBO operator](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/an-slo-for-baremetal.md#standard-slo-behaviors).
+
+### cluster-hosted-net-services-operator details
+
+The CHNSO should:
+
+- Be a new `openshift/cluster-hosted-net-services-operator` project.
+- Publish an image called `cluster-hosted-net-services-operator`.
+- Add a new `ClusterOperator` with an additional
+  `Disabled` status for non on-prem platforms.
+- Implement a controller reconciling a singleton instance representing the desired cluster hosted LB/DNS/VIP configuration.
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource is a platform
+  type other than openstack, baremetal, ovirt or vsphere.
+
+
+#### CRD definition
+
+Below is a possible option for a CRD instance of this operator:
+
+
+```yaml
+apiVersion: clusterHostedNetServices.operator.io/v1alpha1
+kind: HostedNetServices
+metadata:
+  name: cluster
+spec:
+  dns: 
+    nodesResolution: Enabled
+    appsResolution: Enabled
+    apiintIpAddress: 192.168.111.5 
```

Comment:
What is "apiintIpAddress"? The API internal IP?

Reply:
Yep, it's the IP address api-int should point to

Reply:
Use CamlCase.